gensim/models/atmodel.py

```diff
@@ -0,0 +1,1017 @@
+            # First check that corpus is actually a list.
+            assert isinstance(corpus, list), "If serialized == False, all input corpora must be lists."
+            self.corpus.extend(corpus)
+
+
+    def compute_phinorm(self, ids, authors_d, expElogthetad, expElogbetad):
+        """Efficiently computes the normalizing factor in phi."""
+        phinorm = np.zeros(len(ids))
+        expElogtheta_sum = np.zeros(self.num_topics)
+        for a in xrange(len(authors_d)):
+            expElogtheta_sum += expElogthetad[a, :]
+        phinorm = expElogtheta_sum.dot(expElogbetad)
+
+        return phinorm
+
+    def inference(self, chunk, author2doc, doc2author, rhot, collect_sstats=False, chunk_doc_idx=None):
+        """
+        Given a chunk of sparse document vectors, update gamma (parameters
+        controlling the topic weights) for each author corresponding to the
+        documents in the chunk.
+
+        The whole input chunk of document is assumed to fit in RAM; chunking of
+        a large corpus must be done earlier in the pipeline.
+
+        If `collect_sstats` is True, also collect sufficient statistics needed
+        to update the model's topic-word distributions, and return a 2-tuple
+        `(gamma_chunk, sstats)`. Otherwise, return `(gamma_chunk, None)`. 
+        `gamma_cunk` is of shape `len(chunk_authors) x self.num_topics`, where 
+        `chunk_authors` is the number of authors in the documents in the
+        current chunk.
+
+        Avoids computing the `phi` variational parameter directly using the
+        optimization presented in **Lee, Seung: Algorithms for non-negative matrix factorization, NIPS 2001**.
+
+        """
+        try:
+            _ = len(chunk)
+        except:
+            # convert iterators/generators to plain list, so we have len() etc.
+            chunk = list(chunk)
+        if len(chunk) > 1:
+            logger.debug("performing inference on a chunk of %i documents", len(chunk))
+
+        # Initialize the variational distribution q(theta|gamma) for the chunk
+        if collect_sstats:
+            sstats = np.zeros_like(self.expElogbeta)
+        else:
+            sstats = None
+        converged = 0
+
+        # Stack all the computed gammas into this output array.
+        gamma_chunk = np.zeros((0, self.num_topics))
+
+        # Now, for each document d update gamma and phi w.r.t. all authors in those documents.
+        for d, doc in enumerate(chunk):
+            if chunk_doc_idx is not None:
+                doc_no = chunk_doc_idx[d]
+            else:
+                doc_no = d
+            # Get the IDs and counts of all the words in the current document.
+            if doc and not isinstance(doc[0][0], six.integer_types):
+                # make sure the term IDs are ints, otherwise np will get upset
+                ids = [int(id) for id, _ in doc]
+            else:
+                ids = [id for id, _ in doc]
+            cts = np.array([cnt for _, cnt in doc])
```

Comment:
Good idea. I noted that a refactor can be made, and explained a bit about it, in the top of the file. I made it a comment though, rather than in the docstring, because I don't want it showing up in the API reference.